man/get_constant.3
.\" get_constant.3.  The get_constant man page.
.\"
.\" (C) 2008 D. V. Wiebe
.\"
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\"
.\" This file is part of the GetData project.
.\"
.\" Permission is granted to copy, distribute and/or modify this document
.\" under the terms of the GNU Free Documentation License, Version 1.2 or
.\" any later version published by the Free Software Foundation; with no
.\" Invariant Sections, with no Front-Cover Texts, and with no Back-Cover
.\" Texts.  A copy of the license is included in the `COPYING.DOC' file
.\" as part of this distribution.
.\"
.TH get_constant 3 "19 October 2009" "Version 0.6.0" "GETDATA"
.SH NAME
get_constant \(em retrieve a scalar constant from a dirfile database
.SH SYNOPSIS
.B #include <getdata.h>
.HP
.nh
.ad l
.BI "int get_constant(DIRFILE *" dirfile ", const char *" field_code ,
.BI "gd_type_t " return_type ", void *" data_out );
.hy
.ad n
.SH DESCRIPTION
The
.BR get_constant ()
function queries a dirfile(5) database specified by
.I dirfile
for the
.B CONST
scalar
.IR field_code ,
which may contain a representation suffix.
The constant is converted to the data type specified by
.IR return_type ,
and stored in the user-supplied buffer
.IR data_out .

The 
.I dirfile
argument must point to a valid DIRFILE object previously created by a call to
.BR dirfile_open (3).
The argument
.I data_out
must point to a valid memory location of sufficient size to hold a value of the
return type specified.

The 
.I return_type
argument should be one of the following symbols, which indicates the desired
return type of the data:
.RS
.TP 11
.I GD_UINT8
unsigned 8-bit integer
.TP
.I GD_INT8
signed (two's complement) 8-bit integer
.TP
.I GD_UINT16
unsigned 16-bit integer
.TP
.I GD_INT16
signed (two's complement) 16-bit integer
.TP
.I GD_UINT32
unsigned 32-bit integer
.TP
.I GD_INT32
signed (two's complement) 32-bit integer
.TP
.I GD_UINT64
unsigned 64-bit integer
.TP
.I GD_INT64
signed (two's complement) 64-bit integer
.TP
.IR GD_FLOAT32
IEEE-754 standard 32-bit single precision floating point number
.TP
.IR GD_FLOAT64
IEEE-754 standard 64-bit double precision floating point number
.TP
.IR GD_COMPLEX64
C99-conformant 64-bit single precision complex number
.TP
.IR GD_COMPLEX128
C99-conformant 128-bit double precision complex number
.TP
.I GD_NULL
the null type: the database is queried as usual, but no data is returned.
In this case,
.I data_out
is ignored and may be NULL.
.RE

The return type of the constant need not be the same as the type of the data
stored in the database.  Type conversion will be performed as necessary to
return the requested type.  If the
.I field_code
does not indicate a representation, but conversion from a complex value to a
purely real one is required, only the real portion of the requested vector will
be returned.
.SH RETURN VALUE
On success,
.BR get_constant ()
returns zero.  On error, it returns -1 and sets the dirfile error to a non-zero
value.  Possible error values are:
.TP 8
.B GD_E_BAD_CODE
The field specified by
.I field_code
was not found in the database.
.TP
.B GD_E_BAD_DIRFILE
An invalid
.I dirfile
was supplied.
.TP
.B GD_E_BAD_FIELD_TYPE
The supplied
.I field_code
referred to a field of a type other than 
.BR CONST .
The caller should use
.BR getdata (3),
or
.BR get_string (3)
instead.
.TP
.B GD_E_BAD_REPR
The representation suffix specified in
.IR field code ,
or in one of the field codes it uses for input, was invalid.
.TP
.B GD_E_BAD_TYPE
An invalid
.I return_type
was specified.
.TP
.B GD_E_INTERNAL_ERROR
An internal error occurred in the library while trying to perform the task.
This indicates a bug in the library.  Please report the incident to the
maintainer.
.P
The dirfile error may be retrieved by calling
.BR get_error (3).
A descriptive error string for the last error encountered may be obtained from
a call to
.BR get_error_string (3).
.SH SEE ALSO
.BR dirfile (5),
.BR dirfile_open (3),
.BR get_constants (3),
.BR get_error (3),
.BR get_error_string (3),
.BR put_constant (3)


NEWS
New in version 0.6.0rc3:

  License

  * The GetData library and its bindings are now distributed under the GNU
    Lesser Public License, version 2.1 or later.

  Dirfile Changes

  * GetData now support Dirfile Standards Version 7.  Standards Version 7 adds
    complex data types, a signed bitfield (SBIT), a polynomial derived field
    (POLYNOM), and other features detailed below.
  * Field codes can now contain a "representation suffix", which allows
    specifying a complex norm used to convert complex valued data into purely
    real data.
  * The n_fields parameter in a LINCOM specification is now optional.
  * A new encoding scheme which supports LZMA compressed data using the .xz
    container format has been added.
  * Look-up tables may contain a complex y-value.  This prohibits writing to
    the associated LINTERP field.

  Library Changes

  * The field parameters spf (RAW), bitnum and numbits (BIT and SBIT) are now
    internally restricted to 16-bit signed numbers, which should be more than
    enough.  The public API still calls these 'int', but callers should never
    assume this means more than 16-bits.
  * The shift field parameter is now a long int, which ensures at least 32-bits.
  * The dirfile flags, including encoding and endianness symbols are now
    explicitly long ints.  Previously these quantities mixed long int with int
    arbitrarily.  This affects the public API.
  * The parser callback has received a new, API breaking signature, which should
    hopefully reduce the amount of API breakage in the future.  It is also now
    possible for the caller to pass data through GetData to the callback, if
    desired.
  * Computation of derived fields can now handle complex valued inputs, and
    computation will occur in the complex plane when required.
  * The computation of LINCOM fields has been re-written and is now ~20% faster
    for LINCOMs with 2 or 3 input fields.  (No change for LINCOMs of 1 field.)
  * A new open flag, GD_PRETTY_PRINT, tells the library to attempt to write a
    nicer looking format file.  What this specifically means, isn't part of the
    public API and shouldn't be relied on in portable code.
  * BUG FIX: dirfile_madd_spec() and dirfile_add_spec() didn't properly flag
    the fragments they modified as dirty, leading to possible loss of metadata.
  * BUG FIX: calling get_nframes on a bzip2 encoded dirfile would report the
    wrong dirfile length.
  * BUG FIX: LINTERP tables are now closed after loading, resolving a file
    descriptor leak of great age (ie. pre-0.3.0).
  * BUG FIX: When re-writing a format file fragment containing fields with
    CONST scalar field parameters, GetData no longer replaces the CONST field
    codes in the field specification lines with the value of the CONST field.
  * BUG FIX: when moving a field with dirfile_move(), the fragment index in
    the entry objects of metafields defined for the field were not updated.
    (The metafields were placed in the correct fragment when the metadata was
    flushed to disk, despite this.)
  * BUG FIX: Calling getdata() or putdata() on 64-bit architectures in certain
    situations involving a non-zero frameoffset no longer results in bizarrity.

  API Changes

  * This release breaks ABI compatibility and, to a lesser extent, API
    compatibility.  The SOVERSION of the library has been incremented
    accordingly.
  * An array containing the CONST scalar field codes used when non-literal
    parameters are used in the specification of various different fields has
    been added to the gd_entry_t object.  As a side-effect of this,
    dirfile_add(), dirfile_alter_entry(), and dirfile_madd() can be used to
    manipulate non-literal field parameters.
  * A function, get_native_type(), now exists which returns the underlying data
    type of a specified field code.
  * A function, get_framenum(), is now available to perform a "reverse look-up"
    on a field: given a data value, it reports the frame number where the field
    has that value.  The field must be monotonic.
  * A function, dirfile_validate(), now exists which can be used to check the
    validity of a field code before use.
  * Two new error codes have been added: GD_E_DOMAIN, potentially returned by
    get_framenum(), and GD_E_BAD_REPR, indicating an unrecognised representation
    suffix, or an attempt to write to a representation.
  * A new syntax error suberror code has been added: GD_E_FORMAT_LITERAL, which
    indicates a syntax error in a complex literal number.
  * BUG FIX: On success, dirfilename(), dirfile_reference(), and get_reference()
    didn't change the error code to GD_E_OK, resulting in spurious error codes
    from these functions.
  * BUG FIX: dirfile_madd_string() and dirfile_madd_const() now don't fail
    erroneously by reporting GD_E_BAD_CODE.
  * BUG FIX: error strings relating to GD_E_DELETE error codes now report the
    proper error information.
  * BUG FIX: the list returned by get_mstring() is now properly NULL terminated.
  * BUG FIX: the entry->table member now reports the proper path to the LUT.
    (That is, the path stored in the format file.)  Previously it incorrectly
    reported a path relative to the base dirfile directory.

  Legacy API Changes

  * Because they would otherwise be unreported, in the structure returned by
    GetFormat(), POLYNOM entries are reported as LINCOM entries, by discarding
    higher order terms.  Similarly, SBIT entries are reported as BIT entries.

  Bindings

  * Bindings for Python and the Interactive Data Language (IDL) have been
    added.
  * Bindings for the functionality introduced into 0.6.0 have been added.
  * C++: The missing bindings for dirfilename(), get_nvectors(), and
    get_fragment_index() have been added.
  * C++: Empty entry class constructors have been added which allow delayed
    specification of field metadata.
  * C++: The dirfile.h header now includes all other C++ headers, and is now

src/get_constant.c
/* (C) 2008 D. V. Wiebe
 *
 ***************************************************************************
 *
 * This file is part of the GetData project.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * GetData is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with GetData; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
#include "internal.h"

/* this function is little more than a public boilerplate for _GD_DoField */
size_t get_constant(DIRFILE* D, const char *field_code, gd_type_t return_type,
    void *data_out)
{
  size_t n_read;
  gd_entry_t *entry;

  dtrace("%p, \"%s\", 0x%x, %p", D, field_code, return_type, data_out);

  if (D->flags & GD_INVALID) {/* don't crash */
    _GD_SetError(D, GD_E_BAD_DIRFILE, 0, NULL, 0, NULL);
    dreturn("%zi", 0);
    return 0;
  }

  _GD_ClearError(D);

  entry = _GD_GetEntry(D, field_code);
    
  if (D->error != GD_E_OK)
    n_read = 0;
  else if (entry && entry->field_type != GD_CONST_ENTRY) {
    _GD_SetError(D, GD_E_BAD_FIELD_TYPE, GD_E_FIELD_GET, NULL, 0, field_code);
    n_read = 0;
  } else
    n_read = _GD_DoField(D, entry, field_code, 0, 0, 0, 0, return_type,
        data_out);

  dreturn("%zi", n_read);
  return n_read;
}
/* vim: ts=2 sw=2 et tw=80
*/
